aom/src/main/java/com/nedap/archie/adlparser/ADLParser.java

```diff
@@ -70,46 +70,53 @@ public ADLParser(MetaModels models) {
-
-        lexer = new AdlLexer(stream);
-        lexer.addErrorListener(errorListener);
-        parser = new AdlParser(new CommonTokenStream(lexer));
-        parser.addErrorListener(errorListener);
-        tree = parser.adl(); // parse
-
-        ADLListener listener = new ADLListener(errors, metaModels);
-        walker= new ParseTreeWalker();
-        walker.walk(listener, tree);
-        Archetype result = listener.getArchetype();
-        //set some values that are not directly in ODIN or ADL
-        ArchetypeParsePostProcesser.fixArchetype(result);
-
-        if(modelConstraintImposer != null && result.getDefinition() != null) {
-            modelConstraintImposer.imposeConstraints(result.getDefinition());
-        } else if (metaModels != null) {
-            metaModels.selectModel(result);
-            if(metaModels.getSelectedBmmModel() != null) {
-                ModelConstraintImposer imposer = new BMMConstraintImposer(metaModels.getSelectedBmmModel());
-                imposer.setSingleOrMultiple(result.getDefinition());
-            } else if (metaModels.getSelectedModelInfoLookup() != null) {
-                ModelConstraintImposer imposer = new ReflectionConstraintImposer(metaModels.getSelectedModelInfoLookup());
-                imposer.setSingleOrMultiple(result.getDefinition());
+        Archetype result = null;
+        try {
+            lexer = new AdlLexer(stream);
+            lexer.addErrorListener(errorListener);
+            parser = new AdlParser(new CommonTokenStream(lexer));
+            parser.addErrorListener(errorListener);
+            tree = parser.adl(); // parse
+
+            ADLListener listener = new ADLListener(errors, metaModels);
+            walker = new ParseTreeWalker();
+            walker.walk(listener, tree);
+            result = listener.getArchetype();
+            //set some values that are not directly in ODIN or ADL
+            ArchetypeParsePostProcesser.fixArchetype(result);
+
+            if (modelConstraintImposer != null && result.getDefinition() != null) {
+                modelConstraintImposer.imposeConstraints(result.getDefinition());
+            } else if (metaModels != null) {
+                metaModels.selectModel(result);
+                if (metaModels.getSelectedBmmModel() != null) {
+                    ModelConstraintImposer imposer = new BMMConstraintImposer(metaModels.getSelectedBmmModel());
+                    imposer.setSingleOrMultiple(result.getDefinition());
+                } else if (metaModels.getSelectedModelInfoLookup() != null) {
+                    ModelConstraintImposer imposer = new ReflectionConstraintImposer(metaModels.getSelectedModelInfoLookup());
+                    imposer.setSingleOrMultiple(result.getDefinition());
+                }
+            }
+            return result;
+        } finally {
+            if (errors.hasErrors()) {
+                throw new ADLParseException(errors, result);
             }
         }
```

Comment:
true. It is sort of intentional, as it should parse into an object model even if it contains errors, for support of modeling tools such as the visual studio code extension. But perhaps the following is better:

```
try to parse, just throw any exceptions if they happen
if any errors:
   construct archetype if possible, eating any exceptions, then throw ADLParseException
else
   return as normal and don't eat any exceptions
```

Is that better?

Reply:
I think I liked the previous situation better. If eating the exceptions is intended, that's fine. Also in the new situation, errors added to the list in the `convertToAOM` method will be ignored.

Reply:
You're probably right, especially since errors will likely be added to the list when walking the tree, not just when creating the lexer and parser. I think moving the creation of the lexer and parser out of the try could help - there are a couple of cases where ANTLR can throw exceptions there, and those should be thrown instead of our own exception.

It's also possible to add an Exception variable to our own exception, so it no longer hides any exception. Would that be better?

Reply:
Adding the exception makes the code rather complicated. I think it's fine as is after my last commit. It will no longer eat any ANTLR exceptions, and in case of parser errors it will still do the best effort conversion that it used to do. 

Should be safe, and the errors variable should only be used now if you want the warnings - no exception is thrown for those, of course.